bin/git-update-hook

```diff
@@ -22,7 +45,11 @@ tmp=$(mktemp /tmp/git.update.XXXXXX)
 log=$(mktemp /tmp/git.update.log.XXXXXX)
 tree=$(mktemp /tmp/git.diff-tree.XXXXXX)
 
-git diff-tree -r "$2" "$3" > $tree
+FROM="$2"
+if [[ "$2" == $NULL_SHA1 ]]; then
+    FROM=$(git rev-parse $(git rev-list --reverse "$3" --not --branches|head -1)^)
+fi
+git diff-tree -r "$FROM" "$3" > $tree
```

Comment:
I think this will have problems when pushing if `$3` is already pointed to by another branch (which you might do if you're pushing a branch as a backup before rewriting the other, for example.) Perhaps this part of the script could be guarded with:

```shell
if [ -n "$(git branch -a --contains "$3")" ]
then
    echo "$3 is already pointed to by another branch, assuming it's already been checked"
    rm -f $log $tmp $tree
    exit $exit_status
fi
```

?